### PR TITLE
Allowing us access to the API Response

### PR DESCRIPTION
**Pull request by krwoodal**
This will give use the API response for if there is a success or error.

### CODE REVIEW

**Comment by krwoodal** [comment]
@rex-remind101  I think so cause we can still get a 200 but an error frag so it would be helpful to have the request id there as well

**Comment by rex-remind101** [comment]
you can use `didFinishRequest` to inspect the payload and throw an error in those cases.

https://github.com/remind101/AutoGraph/blob/cf2ecd10209aaf200e8d4ce4a89008a46eb4c288/AutoGraph/Request.swift#L25

File: AutoGraph/Errors.swift [hunk 1]
```diff
@@ -36,25 +36,24 @@ public protocol NetworkError: Error {
 public typealias NetworkErrorParser = (_ graphQLError: GraphQLError) -> NetworkError?
 
 public indirect enum AutoGraphError: LocalizedError {
-    case graphQL(errors: [GraphQLError])
+    case graphQL(errors: [GraphQLError], response: HTTPURLResponse?)
     case network(error: Error, statusCode: Int, response: HTTPURLResponse?, underlying: AutoGraphError?)
-    case mapping(error: Error)
-    case refetching(error: Error?)
+    case mapping(error: Error, response: HTTPURLResponse?)
     case typeCoercion(from: Any.Type, to: Any.Type)
```
Comment (rex-remind101) [inline AutoGraph/Errors.swift]: I think we can delete typeCoercion error too

File: AutoGraph/Utilities.swift [hunk 2]
```diff
@@ -23,7 +23,7 @@ extension DataResponse {
                         return nil
                 }
                 
-                return AutoGraphError(graphQLResponseJSON: json, networkErrorParser: nil)
+                return AutoGraphError(graphQLResponseJSON: json, networkErrorParser: nil, response: self.response)
```
Comment (rex-remind101) [inline AutoGraph/Utilities.swift]: We should put the closure at the end of the args list

File: AutoGraphTests/ResponseHandlerTests.swift [hunk 3]
```diff
@@ -57,7 +57,7 @@ class ResponseHandlerTests: XCTestCase {
                 return
             }
             
-            guard case .graphQL(errors: let errors) = error else {
+            guard case .graphQL(errors: let errors, _) = error else {
```
Comment (rex-remind101) [inline AutoGraphTests/ResponseHandlerTests.swift]: Maybe test that we get a response in the right cases?

File: AutoGraphTests/ResponseHandlerTests.swift [hunk 4]
```diff
@@ -141,4 +141,129 @@ class ResponseHandlerTests: XCTestCase {
...
+            }
+            
+             requestId = response?.allHeaderFields["request_id"] as? String
+        }, preMappingHook: { (_, _) in })
+        
+        XCTAssertNotNil(requestId)
+        XCTAssertEqual(requestId, "1234")
+    }
+    
+    func testResponseReturnedInvaildResponse() {
```
Comment (rex-remind101) [inline AutoGraphTests/ResponseHandlerTests.swift]: `testResponseReturnedInvaildResponse` -> `testResponseReturnedFromInvalidResponseError`

File: AutoGraphTests/ResponseHandlerTests.swift [hunk 5]
```diff
@@ -141,4 +141,129 @@ class ResponseHandlerTests: XCTestCase {
...
+            }
+            
+            requestId = response?.allHeaderFields["request_id"] as? String
+        }, preMappingHook: { (_, _) in })
+        
+        XCTAssertNotNil(requestId)
+        XCTAssertEqual(requestId, "1234")
+    }
+    
+    func testResponseReturnedFromMapping() {
```
Comment (rex-remind101) [inline AutoGraphTests/ResponseHandlerTests.swift]: `testResponseReturnedFromMapping` -> `testResponseReturnedFromMappingError`

File: AutoGraphTests/ResponseHandlerTests.swift [hunk 6]
```diff
@@ -141,4 +141,129 @@ class ResponseHandlerTests: XCTestCase {
...
+            }
+            
+            requestId = response?.allHeaderFields["request_id"] as? String
+        }, preMappingHook: { (_, _) in })
+        
+        XCTAssertNotNil(requestId)
+        XCTAssertEqual(requestId, "1234")
+    }
+    
+    func testResponseReturnedGraphQLError() {
```
Comment (rex-remind101) [inline AutoGraphTests/ResponseHandlerTests.swift]: `testResponseReturnedGraphQLError` -> `testResponseReturnedFromGraphQLError`

**Pull request merged**